staging/src/k8s.io/apiextensions-apiserver/test/integration/defaulting_test.go

```diff
@@ -400,6 +408,239 @@ func testDefaulting(t *testing.T, watchCache bool) {
+          unknown: unknown
+  preserve-fullObject:
+    type: object
+    x-kubernetes-embedded-resource: true
+    x-kubernetes-preserve-unknown-fields: true
+    default:
+      apiVersion: foos/v1
+      kind: Foo
+      metadata:
+        name: Bar
+        unknown: unknown
+  preserve-spanning:
+    type: object
+    properties:
+      embedded:
+        type: object
+        x-kubernetes-embedded-resource: true
+        x-kubernetes-preserve-unknown-fields: true
+    default:
+      embedded:
+        apiVersion: foos/v1
+        kind: Foo
+        metadata:
+          name: Bar
+          unknown: unknown
+`
+
+const metaDefaultingFooInstance = `
+kind: Foo
+apiVersion: tests.example.com/v1beta1
+metadata:
+  name: foo
+`
+
+func TestCustomResourceDefaultingOfMetaFields(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CustomResourceDefaulting, true)()
+
+	tearDownFn, apiExtensionClient, dynamicClient, err := fixtures.StartDefaultServerWithClients(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tearDownFn()
+
+	crd := metaDefaultingFixture.DeepCopy()
+	crd.Spec.Versions[0].Schema = &apiextensionsv1.CustomResourceValidation{}
+	if err := yaml.Unmarshal([]byte(metaDefaultingFooV1beta1Schema), &crd.Spec.Versions[0].Schema.OpenAPIV3Schema); err != nil {
+		t.Fatal(err)
+	}
+	crd, err = fixtures.CreateNewV1CustomResourceDefinition(crd, apiExtensionClient, dynamicClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Logf("Creating CR and expecting defaulted, embedded objects, with the unknown ObjectMeta fields pruned")
+	fooClient := dynamicClient.Resource(schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Versions[0].Name, Resource: crd.Spec.Names.Plural})
+
+	tests := []struct {
+		path  []string
+		value interface{}
+	}{
+		{[]string{"fields"}, map[string]interface{}{"metadata": map[string]interface{}{}}},
+		{[]string{"fullMetadata"}, map[string]interface{}{}},
+		{[]string{"fullObject"}, nil},
+		{[]string{"spanning", "embedded"}, nil},
+		{[]string{"preserve-fields"}, map[string]interface{}{"metadata": map[string]interface{}{}}},
+		{[]string{"preserve-fullMetadata"}, map[string]interface{}{}},
+		{[]string{"preserve-fullObject"}, nil},
+		{[]string{"preserve-spanning", "embedded"}, nil},
+	}
+
+	foo := &unstructured.Unstructured{}
+	if err := yaml.Unmarshal([]byte(metaDefaultingFooInstance), &foo.Object); err != nil {
+		t.Fatal(err)
+	}
+	for _, tst := range tests {
+		if tst.value != nil {
+			if err := unstructured.SetNestedField(foo.Object, tst.value, tst.path...); err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+	foo, err = fooClient.Create(foo, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Unable to create CR: %v", err)
+	}
+	t.Logf("CR created: %#v", foo.UnstructuredContent())
+
+	for _, tst := range tests {
+		t.Run(strings.Join(tst.path, "."), func(t *testing.T) {
+			obj, found, err := unstructured.NestedMap(foo.Object, tst.path...)
```

Comment:
this tests that defaults of unknown fields didn't survive defaulting->serialization->etcd->decoding->pruning, but it's not clear that the pruning happened before data got persisted to etcd (which is what we really care about)

can we use the storage.NewEtcdObjectReader (from the conversion webhook tests) to pull data out of etcd for verification?

Reply:
yep, I noticed that as well. Let me check about etcd object reader.

Reply:
done. And verified that without the fixed default pruning algorithm it breaks :+1: